### PR TITLE
Workload attestation

### PR DESCRIPTION
**Pull request by waynz0r**
## Description

Please provide a meaningful description of what this change will do, or is for. Bonus points for including links to
related issues, other PRs, or technical references.

Note that by _not_ including a description, you are asking reviewers to do extra work to understand the context of this
change, which may lead to your PR taking much longer to review, or result in it not being reviewed at all.

## Type of Change

- [ ] Bug Fix
- [ ] New Feature
- [ ] Breaking Change
- [ ] Refactor
- [ ] Documentation
- [ ] Other (please describe)

## Checklist

- [ ] I have read the [contributing guidelines](/CONTRIBUTING.md)
- [ ] Existing issues have been referenced (where applicable)
- [ ] I have verified this change is not present in other open pull requests
- [ ] Functionality is documented
- [ ] All code style checks pass
- [ ] New code contribution is covered by automated tests
- [ ] All new and existing tests pass

### CODE REVIEW

File: commands.c [hunk 1]
```diff
@@ -81,14 +82,69 @@ command_answer *send_command(char *name, char *data, task_context *context)
...
+command_answer *send_attest_command(direction direction, struct sock *s, u16 port)
+{
+    int iplen = 16;
+    char ipformat[] = "%pI4";
+
+    if (s->sk_family == AF_INET6)
+    {
+        iplen = 46;
+        strcpy(ipformat, "%pI6");
+    }
```
Comment (baluchicken) [inline commands.c:94]: I would avoid the strcpy function, instead I would use `cont char* ipformat`. and simply update the pointer value inside the if.

File: commands.c [hunk 2]
```diff
@@ -81,14 +82,69 @@ command_answer *send_command(char *name, char *data, task_context *context)
...
+    else
+    {
+        snprintf(source_ip, sizeof(source_ip), ipformat, &s->sk_rcv_saddr);
+        snprintf(destination_ip, sizeof(destination_ip), ipformat, &s->sk_daddr);
+        source_port = s->sk_num;
+        destination_port = port;
+    }
+
+    JSON_Value *root_value = json_value_init_object();
+    JSON_Object *root_object = json_value_get_object(root_value);
```
Comment (baluchicken) [inline commands.c:117]: We must check if these return values are NULLs or not.

File: commands.c [hunk 3]
```diff
@@ -81,14 +82,69 @@ command_answer *send_command(char *name, char *data, task_context *context)
...
+    json_object_set_number(root_object, "source_port", source_port);
+    json_object_set_string(root_object, "destination_ip", destination_ip);
+    json_object_set_number(root_object, "destination_port", destination_port);
+
+    char *serialized_string = json_serialize_to_string(root_value);
+
+    command_answer *answer = send_command("attest", serialized_string, get_task_context());
+
+    json_free_serialized_string(serialized_string);
+    json_value_free(root_value);
```
Comment (baluchicken) [inline commands.c:130]: This json_value_free function name is misleading for me. I would add a command that this function also frees the root_object field.

File: commands.c [hunk 4]
```diff
@@ -81,14 +82,69 @@ command_answer *send_command(char *name, char *data, task_context *context)
...
+
+    return answer;
+}
+
 command_answer *send_accept_command(u16 port)
 {
     JSON_Value *root_value = json_value_init_object();
```
Comment (baluchicken) [inline commands.c:137]: The NULL check should apply here as well.

File: commands.c
Comment (baluchicken) [inline commands.c:1]: The whole file lacks the NULL check for 
```
JSON_Value *root_value = json_value_init_object();
 JSON_Object *root_object = json_value_get_object(root_value);
  ```

File: device_driver.c [hunk 5]
```diff
@@ -206,14 +206,14 @@ wasm_vm_result load_module(char *name, char *code, unsigned length, char *entryp
...
+int parse_json_from_buffer(const char *data)
 {
     int status = SUCCESS;
     JSON_Value *json = NULL;
 
-    if (device_buffer_size)
+    if (data)
     {
-        json = json_parse_string(device_buffer);
+        json = json_parse_string(data);
```
Comment (baluchicken) [inline device_driver.c:216]: This can be NULL, we must check this to avoid NULL pointer exception.

File: device_driver.c [hunk 6]
```diff
@@ -225,11 +225,13 @@ int parse_json_from_buffer(void)
             printk("nasp: loading module: %s", name);
 
             char *code = json_object_get_string(root, "code");
-            int length = base64_decode(device_buffer, DEVICE_BUFFER_SIZE, code, strlen(code));
+            char *decoded = kzalloc(strlen(code) * 2, GFP_KERNEL);
```
Comment (baluchicken) [inline device_driver.c:228]: This field leaks memory if no error happens. We should kfree even when the load module succeeds.

File: device_driver.c [hunk 7]
```diff
@@ -260,22 +263,32 @@ int parse_json_from_buffer(void)
...
+        else if (strcmp("load_rules", command) == 0)
         {
-            char *base64_id = json_object_get_string(root, "id");
-
-            printk("nasp: command answer parsing, id: %s", base64_id);
-
-            char command_id[UUID_SIZE * 2];
-            int length = base64_decode(command_id, UUID_SIZE * 2, base64_id, strlen(base64_id));
+            char *code = json_object_get_string(root, "code");
+            char *decoded = kzalloc(strlen(code) * 2, GFP_KERNEL);
```
Comment (baluchicken) [inline device_driver.c:269]: We should kfree decoded in case of length <0

File: device_driver.c [hunk 8]
```diff
@@ -260,22 +263,32 @@ int parse_json_from_buffer(void)
...
-                FATAL("base64_decode of id failed");
+                FATAL("base64_decode failed");
                 status = -1;
                 goto cleanup;
             }
 
-            struct command *cmd = lookup_in_flight_command(command_id);
+            printk("nasp: loading rules");
+
+            load_opa_data(strdup(decoded));
```
Comment (baluchicken) [inline device_driver.c:280]: Why do we need to strdup the decoded data here?

File: device_driver.c [hunk 9]
```diff
@@ -316,6 +329,8 @@ int parse_json_from_buffer(void)
         json_value_free(json);
     }
 
+    kfree(data);
```
Comment (baluchicken) [inline device_driver.c:332]: I think this is a bad idea to free data that comes as a function parameter.

File: device_driver.c [hunk 10]
```diff
@@ -339,8 +354,8 @@ static int device_release(struct inode *inode, struct file *file)
 
 static int write_command_to_buffer(char *buffer, size_t buffer_size, struct command *cmd)
 {
-    char uuid[UUID_SIZE * 2];
-    int length = base64_encode(uuid, UUID_SIZE * 2, cmd->uuid.b, UUID_SIZE);
+    char uuid[UUID_SIZE + 21];
```
Comment (baluchicken) [inline device_driver.c:357]: What is 21? Please create a constant with a meaningful name.

File: device_driver.c [hunk 11]
```diff
@@ -464,11 +492,10 @@ static ssize_t device_write(struct file *file, const char *buffer, size_t length
         }
 
         // parse the json
-        int status = parse_json_from_buffer();
+        int status = parse_json_from_buffer(strdup(device_buffer));
```
Comment (baluchicken) [inline device_driver.c:495]: If we remove the kfree(data) from the parse_json_from_buffer function this strdup is not needed.

File: device_driver.c [hunk 12]
```diff
@@ -464,11 +492,10 @@ static ssize_t device_write(struct file *file, const char *buffer, size_t length
...
         if (status != 0)
         {
             printk(KERN_ERR "nasp: parse_json_from_buffer failed: %d", status);
-            return -1;
```
Comment (baluchicken) [inline device_driver.c:471]: If I understand it correctly. Removing this return here means we are unable to parse the JSON from the buffer but it is fine we are moving forward and removing the failed data from the buffer and moving forward. Is this what is happening here?

**Review by baluchicken — CHANGES_REQUESTED** [review]

**Review by baluchicken — CHANGES_REQUESTED** [review]

File: opa.c [hunk 13]
```diff
@@ -147,29 +175,162 @@ static int parse_opa_eval_result(char *json)
...
+                if (i < json_array_get_count(dns) - 1)
+                {
+                    strcat(ret.dns, ",");
+                }
+            }
+        }
+
+        if (!ret.uri)
+        {
+            ret.uri = strndup("spiffe://unspecified", 20);
```
[the remark this answers is not in the document]
Comment (baluchicken) [reply opa.c:322]: After further investigation, I saw why the strndup needed.

File: device_driver.c (hunk 12 above)
Comment (baluchicken) [reply device_driver.c:471]: After further discussion returning -1 is also a bad idea. It put the agent into an unrecoverable state.

File: socket.c [hunk 14]
```diff
@@ -1055,35 +1025,42 @@ int nasp_connect(struct sock *sk, struct sockaddr *uaddr, int addr_len)
...
-			goto error;
-		}
+	// send attest command
+	command_answer *answer = send_attest_command(OUTPUT, sk, port);
 
-		memcpy(sc->rsa_priv, rsa_priv, sizeof *sc->rsa_priv);
-		memcpy(sc->rsa_pub, rsa_pub, sizeof *sc->rsa_pub);
+	if (answer->error)
+	{
+		pr_err("nasp: connect failed to send attest command: %s", answer->error);
```
Comment (baluchicken) [inline socket.c:1047]: Same as above, shouldn't we goto error in this case?

File: tls.c [hunk 15]
```diff
@@ -61,9 +61,38 @@ xwc_end_chain(const br_x509_class **ctx)
...
+                }
+            }
+        }
+    }
+
+    if (!allowed)
+    {
+        pr_info("nasp: peer certificate is not allowed!");
+
+        return 62;
```
Comment (baluchicken) [inline tls.c:95]: What is 62?

File: socket.c [hunk 16]
```diff
@@ -930,37 +897,42 @@ struct sock *nasp_accept(struct sock *sk, int flags, int *err, bool kern)
...
-			goto error;
-		}
+	// send attest command
+	command_answer *answer = send_attest_command(INPUT, client, port);
 
-		memcpy(sc->rsa_priv, rsa_priv, sizeof *sc->rsa_priv);
-		memcpy(sc->rsa_pub, rsa_pub, sizeof *sc->rsa_pub);
+	if (answer->error)
+	{
+		pr_err("nasp: accept failed to send attest command: %s", answer->error);
```
[the remark this answers is not in the document]
Comment (baluchicken) [reply socket.c:919]: This is a design decision made by us to fall through without mtls when the attestation failes for some reason.

File: socket.c (hunk 14 above)
Comment (baluchicken) [reply socket.c:1047]: This is a design decision made by us to fall through without mtls when the attestation fails for some reason.

File: tls.c (hunk 15 above)
Comment (baluchicken) [reply tls.c:95]: This is an error code:
/** \brief X.509 status: chain could not be linked to a trust anchor. */
#define BR_ERR_X509_NOT_TRUSTED           62

**Review by baluchicken — DISMISSED** [review]

**Review by bonifaido — DISMISSED** [review]

**Review by bonifaido — APPROVED** [review]

**Pull request merged**